### PR TITLE
Draft HTTP Content Negotiation Header Fields

### PR DESCRIPTION
**Pull request by co-category**
This PR adds

- common media types
- content negotiation header fields (see [here](https://tools.ietf.org/html/rfc7231#section-5.3))

### CODE REVIEW

File: core/src/main/scala/zio/http/model/Charset.scala [hunk 1]
```diff
@@ -18,25 +18,18 @@
...
 
-import java.nio.charset.{ StandardCharsets, Charset => JCharset }
+import Charset._
 
-import scala.util.Try
-
-final case class Charset private (value: String) extends AnyVal {
-  override def toString: String = value
-}
+final case class Charset(value: CharsetType, param: Map[String, String] = Map.empty)
```
Comment (co-category) [inline core/src/main/scala/zio/http/model/Charset.scala]: @ioleo I changed the definition of this type here. The general pattern which I noticed regarding header fields is one where there is some type of value and then optionally associated to that value are some parameters and maybe even some extra extension parameters. This encoding allows for that and so can then be used to defined the `Accept-Charset` content negotiation header field.

File: core/src/main/scala/zio/http/model/Language.scala [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+/*
+  Language tags used by Content-Language and Accept-Language header fields.
+ */
+final case class Language(value: String, param: Map[String, String] = Map.empty)
+
+object Language {
+
+  val ab = Language("Abkhazian")
+  val aa = Language("Afar")
```
Comment (co-category) [inline core/src/main/scala/zio/http/model/Language.scala]: I could make these country-language identifiers uppercase - not sure which convention the zio team follows in this regard.

**Comment by co-category** [comment]
See [this](https://github.com/zio/zio-http/pull/17#discussion_r346613331)

File: core/src/main/scala/zio/http/model/Charset.scala (hunk 1 above)
Comment (saraiva132) [reply core/src/main/scala/zio/http/model/Charset.scala]: I do not quite understand this. A charset by itself is only a string that should respect the standards (i.e. RFC 2278) and only that. We can then use these charsets to build a header Accept-Charset which can, as you said, have params associated with it. By doing this we have lost the actual ability to transform a Charset to its string representation (which was leveraging the java nio charset implementation). Perhaps there is something I am missing here,

Comment (co-category) [reply core/src/main/scala/zio/http/model/Charset.scala]: Yes, you are right @saraiva132 . I will add back the ability to transform a Charset to its string representation. The other thing I was thinking is to get the complete set of charsets supported in Java 8.

Comment (saraiva132) [reply core/src/main/scala/zio/http/model/Charset.scala]: I did that myself. The charset there are the ones that are guaranteed to be supported. Always. For additional ones I was not sure whether it was worth adding then so I just left smart constructors to build it from charset or string ( with validation), again, making use of Java charset charset validation

Comment (co-category) [reply core/src/main/scala/zio/http/model/Charset.scala]: kl - I will it change it back.

Comment (co-category) [reply core/src/main/scala/zio/http/model/Charset.scala]: This has been restored as part #24 (I'm going to close this PR) . However, I can't see a good way to associate the quality value to the charset now. Any ideas?

Comment (saraiva132) [reply core/src/main/scala/zio/http/model/Charset.scala]: Is quality value a param ? Could you further elaborate ?

File: core/src/main/scala/zio/http/model/Language.scala (hunk 2 above)
Comment (saraiva132) [reply core/src/main/scala/zio/http/model/Language.scala]: Do we have a standard? @ioleo

**Comment by co-category** [comment]
closed as this is handled by #24

**Pull request closed without merging**